### PR TITLE
Add `torch.linalg.matrix_sqrth` for Symmetric/Hermitian Positive-Definite Matrices

### LINKED ISSUE #25481: [feature request] symmetric matrix square root
This is needed when incorporating curvature information into optimization

There's PyTorch implementation of symmetric matrix square root op [here](https://github.com/msubhransu/matrix-sqrt) but they use PyTorch for backward pass only, and use scipy for forward pass

I've hacked something together below by borrowing logic to detect 0 eigenvalues from scipy.linalg.pinvh

```
 def symsqrt(a, cond=None, return_rank=False):
    """Computes the symmetric square root of a positive definite matrix"""

    s, u = torch.symeig(a, eigenvectors=True)
    cond_dict = {torch.float32: 1e3 * 1.1920929e-07, torch.float64: 1E6 * 2.220446049250313e-16}

    if cond in [None, -1]:
        cond = cond_dict[a.dtype]

    above_cutoff = (abs(s) > cond * torch.max(abs(s)))

    psigma_diag = torch.sqrt(s[above_cutoff])
    u = u[:, above_cutoff]

    B = u @ torch.diag(psigma_diag) @ u.t()
    if return_rank:
        return B, len(psigma_diag)
    else:
        return B

def symsqrt_test():
    mat = torch.reshape(torch.arange(9) + 1, (3, 3)).float() + torch.eye(3) * 5
    mat = mat + mat.t()  # make symmetric
    smat = u.symsqrt(mat)
    u.check_close(mat, smat @ smat.t())
    u.check_close(mat, smat @ smat)

    def randomly_rotate(X):
        """Randomly rotate d,n data matrix X"""
        d, n = X.shape
        z = torch.randn((d, d), dtype=X.dtype)
        q, r = torch.qr(z)
        d = torch.diag(r)
        ph = d / abs(d)
        rot_mat = q * ph
        return rot_mat @ X

    n = 20
    d = 10
    X = torch.randn((d, n))

    # embed in a larger space
    X = torch.cat([X, torch.zeros_like(X)])
    X = randomly_rotate(X)
    cov = X.t() @ X
    sqrt, rank = symsqrt(cov, return_rank=True)
    assert rank == d
    assert torch.allclose(sqrt @ sqrt, cov, atol=1e-5)
```

### PR DESCRIPTION
**Pull request by colalb1**
Adds `torch.linalg.matrix_sqrt`, the principal square root of a symmetric positive-definite matrix, computed from the eigendecomposition as $X = Q  \text{diag}(sqrt(\lambda)) Q^H$. Supports float/double/cfloat/cdouble and batches.

The backward is a custom first-order Loewner differential with denominator $\sqrt{\lambda_i} + \sqrt{\lambda_j}$, so it stays numerically stable at repeated eigenvalues. Since the Loewner matrix is real-symmetric and $Q$ unitary, the same operator serves both the VJP and the JVP. Higher-order derivatives go through eigh and are not stable at degeneracies. Indefinite input raises rather than silently projecting since the real square root of such a matrix is complex.

Beyond the kernel and backward, the operation is fully wired for the eager/compile/functorch stacks: meta registration, functorch vmap batch rule, Dynamo in-graph binding, CPU autocast (fp32), MPS xfail markers, OpInfo, docs, and dedicated tests.

All tests pass on CPU (MPS cases xfail as expected, CUDA not exercised locally). A `torch.compile` forward+backward smoke check on a PD input traces in-graph (fullgraph=True) and matches eager.

Closes #25481

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Related:
- https://github.com/pytorch/pytorch/issues/25481
- https://github.com/pytorch/pytorch/issues/9983

**Comment by colalb1** [comment]
I will make the PR split for the `native_funtions.yaml` soon

**Review by lezcano — COMMENTED** [review]
What do you gain from having this in PyTorch over just implementing it with a custom function with the backward? The perf is going to be the same...

**Comment by colalb1** [comment]
> What do you gain from having this in PyTorch over just implementing it with a custom function with the backward? The perf is going to be the same...

@lezcano 

Yes, the performance will be the same as `eigh` and Daleckii–Krein.

One of the main benefits is that the operation composes with `torch.compile` without a graph break, vmap, meta, autocast, etc. for "free." A user-defined `autograd.Function` does not get any of these without significant per-feature work.

Less relevant, but this also prevents the issue of running autograd through a raw `eigh` that gives the potential for overflow due to tightly clustered eigenvalues `1 / (\lambda_i - \lambda_j)`. Since the implemented operation uses the Loewner form `1 / (\sqrt{\lambda_i} - \sqrt{\lambda_j})`, this stays first-order stable. 

I suppose one could just write the custom function with the backward, but this `sqrt` discrepancy could be an easy miss; we also get OpInfo coverage with this feature.

Let me know what you think. Thank you.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Comment by lezcano** [comment]
I think this belongs to an external library implementing this and other matrix functions via `autograd.Function` + a helpers to support `vmap` and all the other bells and whistles if necessary rather than in core, as adding it to core would not bring any perf over an external implementation.
If there is an implementation in cusolver / magma + BLAS for CPU, then we would certainly accept a PR wiring these through a linalg API tho.

**Comment by colalb1** [comment]
@jansel I just pushed changes to fix the breakages.

**Comment by colalb1** [comment]
@lezcano That makes sense. 

Going off of what you mentioned about a method wired through a linalg API, would a Schur-based `sqrtm` (like [that of SciPy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.sqrtm.html)) wired through LAPACK be something you'd be interested in? This would both close the generality gap to SciPy and include triangular recurrence + Schur factorization through LAPACK kernels (improved performance), which a user could not replicate easily.

If we were to move forward with this, we would have to add a couple of bindings to ATen (`gees`, `trsyl`; via claude).

Before I make changes, is there a particular direction you'd prefer to take this? I'd be happy to make the necessary changes.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -4327,6 +4327,30 @@ Tensor linalg_matrix_exp_differential(
...
+  if (!grad.defined()) {
+    return {};
+  }
+  at::NoTF32Guard disable_tf32;
+  auto [eigvals, eigvecs] = at::linalg_eigh(self);
+  auto sqrt_eigvals = eigvals.clamp_min(0).sqrt();
+  auto denom = sqrt_eigvals.unsqueeze(-1) + sqrt_eigvals.unsqueeze(-2);
+  auto grad_sym = 0.5 * (grad + grad.mH());
+  auto inner =
+      at::matmul(at::matmul(eigvecs.mH(), grad_sym), eigvecs).div(denom);
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp:4351]: Can you invert the denom can change ti multiply to speed it up a bit? Or is that too numerically unstable?

File: torch/testing/_internal/opinfo/definitions/linalg.py [hunk 2]
```diff
@@ -1213,6 +1225,25 @@ def make_input():
         sample_inputs_func=sample_inputs_diagonal_diag_embed,
         error_inputs_func=error_inputs_diagonal_diag_embed,
     ),
+    OpInfo(
+        "linalg.matrix_sqrt",
+        aten_name="linalg_matrix_sqrt",
+        dtypes=floating_and_complex_types(),
+        sample_inputs_func=sample_inputs_linalg_matrix_sqrt,
+        gradcheck_wrapper=gradcheck_wrapper_hermitian_input,
+        check_batched_grad=False,
```
Comment (Skylion007) [inline torch/testing/_internal/opinfo/definitions/linalg.py]: Why not check batched grad, should be easy to add? DTensor should be easy too?

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp:4351]: Numerical stability is not an issue since the input is positive-definite, meaning the reciprocal is well-conditioned.

Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp:4351]: Though numerical stability would be fine with a multiply, my concern is that replacing one kernel division with a reciprocal and a multiply would cause too much overhead with two full passes over the matrix as well as an additional kernel launch.

To my knowledge, the precomputed reciprocal would only save time/compute when `1/denom` is used across many multiplies. Thus, when there is only one use, there is nothing to amortize.

Do you think it is fine to leave the function as-is, or would you prefer a change?

File: torch/testing/_internal/opinfo/definitions/linalg.py (hunk 2 above)
Comment (colalb1) [reply torch/testing/_internal/opinfo/definitions/linalg.py]: Will add batched grad.

Comment (colalb1) [reply torch/testing/_internal/opinfo/definitions/linalg.py]: `matrix_sqrt` is a full-matrix computation on the last two dimensions, meaning its closest analogs are `linalg.matrix_power` and `matrix_exp`, both of which are xfailed in `dtensor_fails_no_strategy`. `matrix_exp` has a sharding strategy but is kept on the xfail list. Thus, I don't believe strategy registration is enough for DTensor OpInfo.

My preference is to have this as a separate, larger change such that this operation + `matrix_exp`/`matrix_power` match.

Edit: I added a comment for the xfail on DTensor.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:4351]: isn't `div(a,b)` implemented as `a*1/b` in cuda?

Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp:4351]: @lezcano To my understanding, yes, at the hardware level what you've specified is what the operation reduces to. The discrepancy between doing `.div()` and `.reciprocal().mul()` is that the fusion happens automatically in the `.div()` kernel. There also may be some level of rounding guarantee that is lost in bypassing the fusion with these two operations.

Comment (colalb1) [reply torch/csrc/autograd/FunctionsManual.cpp:4351]: I read into this more, and it's a double-rounding issue rather than fusion. `a.div(b)` is implemented as a single correctly-rounded operation, giving `round(a/b)`. `b.reciprocal().mul(a)` computes the same mathematical quantity in two separately rounded steps `round(a * round(1/b))`, so the results will not be identical.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:4351]: I know, I was just saying that Skylion's review was not relevant here.

**Comment by lezcano** [comment]
Ok, after thinking about it, I think this makes sense.
We still need to fix CI and then please follow-up with a PR adding LAPACK's API (for symmetric / Hermitian matrices). I agree that claude (or codex...) can do a very good job implementing this as there are a ton of examples of this in the repo. We should benchmark it against the current implementation both in accuracy and perf, as LAPACK's algo is generic so not sure how it fares against our eigh implementation.

A bit of bikeshedding here tho. A better name for this function would probably be `matrix_sqrth` as it acts on Hermitian matrices (same as `eigh`).

**Comment by colalb1** [comment]
@lezcano Sounds good. I will change the function name to `matrix_sqrth`, as suggested.

My understanding is that the remaining 5 CI failures are not related to this PR (first pytorch-bot message).

Just for clarity before proceeding, the remaining steps are as follows:
- Change operation name to `matrix_sqrth`
- Review + merge
- Create branch with LAPACK version
- Benchmark `eigh` against LAPACK version
- Merge LAPACK version (conditional on performance)

Is this what you had in mind?

If there's anything I missed, please let me know.

Edit: Renaming update is complete.

**Comment by lezcano** [comment]
this sounds about right, yeah.
You can rebase on top of `viable/strict` andf ix the lint? This should get rid of the flakes.

I will do a proper review tomorrow.

**Comment by colalb1** [comment]
@lezcano I rebased on `viable/strict`, fixing linting violations in `torch/_meta_registrations.py` and `torch/csrc/autograd/FunctionsManual.cpp`. The remaining Pyrefly linting errors are pre-existing in `torch/cuda/graphs.py`.

**Review by lezcano — COMMENTED** [review]
you mistakenly checked in `kineto`, can you remove that?

**Comment by colalb1** [comment]
@lezcano Removed kineto changes.

**Comment by colalb1** [comment]
@lezcano @Skylion007 
 Hello.

The `mobilenetv2_100` CI failure (eager_two_runs_differ on `blocks.1.0.bn1.bias.grad`) has shown up on two separate runs. After looking at the diff, there's no logical path from any of my changes (that I'm aware of) to BatchNorm gradient non-determinism in `MobileNetV2`. 

Could this be a non-determinism/flaky runner issue? If you know if this is related to my change, please let me know. Thank you

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 3]
```diff
@@ -2807,6 +2808,49 @@ Tensor matrix_exp(const Tensor& a) {
...
+  // pass this check and clamp_min below rounds them up to zero. The reduction
+  // forces one host-device sync per call (eigvals can be empty for a zero-sized
+  // batch, where there is nothing to validate).
+  if (eigvals.sym_numel() != 0) {
+    auto min_eigval = eigvals.min().item<double>();
+    auto max_abs_eigval = eigvals.abs().max().item<double>();
+    auto eps = eigvals.scalar_type() == at::kFloat
+        ? std::numeric_limits<float>::epsilon()
+        : std::numeric_limits<double>::epsilon();
+    auto tol = max_abs_eigval * eps * static_cast<double>(a.size(-1));
```
Comment (lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp]: Remove all this as this will create an unneeded synchronsiation point.

File: test/test_linalg.py [hunk 4]
```diff
@@ -7635,6 +7635,76 @@ def test_linalg_matrix_exp_no_warnings(self, device, dtype):
...
+    @dtypes(torch.float, torch.cfloat)
+    def test_linalg_matrix_sqrth_errors(self, device, dtype):
+        with self.assertRaisesRegex(RuntimeError, "must be batches of square matrices"):
+            torch.linalg.matrix_sqrth(torch.randn(2, 3, device=device, dtype=dtype))
+
+        # The real square root of an indefinite matrix is complex, so a materially
+        # negative eigenvalue is rejected rather than silently projected away.
+        indefinite = torch.diag(torch.tensor([1.0, -2.0], device=device)).to(dtype)
+        with self.assertRaisesRegex(RuntimeError, "positive semi-definite"):
+            torch.linalg.matrix_sqrth(indefinite)
```
Comment (lezcano) [inline test/test_linalg.py]: and remove this. The fact that it's PSD (or semidefinite) is a precondition for the function. We don't check it, the same way we don't check that it's symmetric.

File: torch/linalg/__init__.py [hunk 5]
```diff
@@ -2215,6 +2215,65 @@
...
+Letting :math:`\mathbb{K}` be :math:`\mathbb{R}` or :math:`\mathbb{C}`,
+for a symmetric (resp. Hermitian) positive-definite matrix :math:`A \in \mathbb{K}^{n \times n}`,
+this function returns the unique symmetric (resp. Hermitian) positive-definite matrix
+:math:`X \in \mathbb{K}^{n \times n}` such that
+
+.. math::
+    XX = A.
+
+The result is computed from the eigendecomposition :math:`A = Q \operatorname{diag}(\lambda) Q^H`
+as :math:`X = Q \operatorname{diag}(\sqrt{\lambda}) Q^H`.
```
Comment (lezcano) [inline torch/linalg/__init__.py]: no need to say this as this may or may not be the case in the future.

File: torch/linalg/__init__.py [hunk 6]
```diff
@@ -2215,6 +2215,65 @@
...
+Supports input of float, double, cfloat and cdouble dtypes.
+Also supports batches of matrices, and if :attr:`A` is a batch of matrices then
+the output has the same batch dimensions.
+
+.. note:: Only the lower triangular part of :attr:`A` is used in the computation, and
+          :attr:`A` is assumed to be symmetric (resp. Hermitian). See :func:`torch.linalg.eigh`.
+
+.. note:: :attr:`A` must be positive semi-definite. The real square root of a matrix with
+          a negative eigenvalue is complex, so this function raises a runtime error when
+          :attr:`A` has a materially negative eigenvalue.
```
Comment (lezcano) [inline torch/linalg/__init__.py]: remove as this is already implicit in the definition

File: torch/linalg/__init__.py [hunk 7]
```diff
@@ -2215,6 +2215,65 @@
...
+
+.. note:: The first-order gradient is numerically stable even when :attr:`A` has repeated
+          eigenvalues, unlike differentiating through :func:`torch.linalg.eigh`, but requires
+          :attr:`A` to be positive-definite. Higher-order derivatives differentiate through
+          :func:`torch.linalg.eigh` and are not stable at repeated eigenvalues.
```
Comment (lezcano) [inline torch/linalg/__init__.py]: no need to advertise this.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 3 above)
Comment (colalb1) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Addressed.

File: test/test_linalg.py (hunk 4 above)
Comment (colalb1) [reply test/test_linalg.py]: Addressed.

File: torch/linalg/__init__.py (hunk 5 above)
Comment (colalb1) [reply torch/linalg/__init__.py]: Addressed. Also removed the comment about the function raising a runtime error for a negative eigenvalue, as this check was removed (see previous comment).

File: torch/linalg/__init__.py (hunk 7 above)
Comment (colalb1) [reply torch/linalg/__init__.py]: Fixed.

File: torch/linalg/__init__.py (hunk 6 above)
Comment (colalb1) [reply torch/linalg/__init__.py]: Removed.

**Review by lezcano — APPROVED** [review]

**Comment by colalb1** [comment]
@lezcano Thanks for the review!

Could you re-run CI perchance? The failures seem to be blocking the merge; I don't believe they're related to the code change.

Thank you.
